amacc.c

```diff
@@ -734,24 +740,12 @@ void expr(int lev)
         case MulAssign:
         case DivAssign:
         case ModAssign:
-            otk = tk;
-            *--n=';'; *--n = ty = id->type; *--n = Load;
-            sz = (ty = t) >= PTR2 ? sizeof(int) :
-                                    ty >= PTR ? tsize[ty - PTR] : 1;
-            next(); c = n; expr(otk);
-            if (*n == Num) n[1] *= sz;
-            *--n = (int) c;
-            if (otk < ShlAssign) {
-                *--n = Or + (otk - OrAssign);
-            } else {
-                *--n = Shl + (otk - ShlAssign);
-                // Compound-op bypasses literal const optimizations
-                if (otk == DivAssign) ef_getidx("__aeabi_idiv");
-                if (otk == ModAssign) ef_getidx("__aeabi_idivmod");
-            }
+            *--n=';'; *--n = t; *--n = Load;
+            if (tk < ShlAssign) tk = Or + (tk - OrAssign);
+            else tk = Shl + (tk - ShlAssign);
+            tk |= REENTRANT; compound = 1; expr(Assign);
```

Comment:
I wonder if compound-op bypasses literal const optimizations still make sense.

Reply:
The old way was taking shortcuts while this technique is a "full parse" using the full standard AST rules. It is "complete" while the old way skipped steps.  scripts/disasm can be used to look at the assembly code, so that you can verify.

Reply:
int main()
{
   int *je = 0;
   je += 4 - 2;
   return 0;
}
% amacc -si test.c
ENT 1
// int *je = 0
LEA je (-1)
PSH
IMM 0
SI
// je += 4 - 2
LEA je (-1)
PSH
LI
PSH
IMM 8  // 2 * sizeof (int)
ADD
SI

Reply:
int main()
{
   int n = 5;
   n *= 4 - 2; // should be n = n * (4 - 2), not n = (n * 4) -2
   return 0;
}
% amacc -si test.c
ENT 1
LEA n (-1)
PSH
IMM 5
SI
// n *= 4 - 2;
LEA n (-1)
PSH
LI
PSH
IMM 2 // correct!
MUL
SI